### ISSUE TITLE
2.2.0 Throws Exception for class with custom metaclass

### ISSUE THREAD

**Issue by b-phi** [tox-dev/sphinx-autodoc-typehints#462]
> Handler <function process_docstring at 0x7f12ae4c8b80> for event 'autodoc-process-docstring' threw an exception (exception: unhashable type: '_SchemaMeta')

I've including the relevant skeleton of the classes below. 

```python
class _SchemaMeta(type):
    def __new__(mcs, *args: Any, **kwargs: Any) -> '_SchemaMeta':
        # some logic
        return schema_cls

    def __eq__(cls, other: Any) -> bool:
        ...

class Schema(metaclass=_SchemaMeta):
    ...
```

Elsewhere in the code, we have this which causes the error
```python
class Foo:
    def do_something(self, schema: Schema) -> None:
        """
        Args:
            schema: Some schema.
        """
        ...
```

Edit: I've pinned to 2.1.1 for now, which doesn't throw an error in this case.

**Comment by hoodmane** [comment]
@tikuma-lsuhsc seems likely a regression caused by #459.

**Comment by tikuma-lsuhsc** [comment]
Yep, I did not account for a string annotation as I wasn't even aware such is even an option. I'll look into it tonight.

**Comment by tikuma-lsuhsc** [comment]
@b-phi - I tested all three class def's of yours but none produced the error that you reported.

Built with `sphinxcontrib-restbuilder` (for convenience as I'm using it for my current project), your code snippet produced the following without any error or warning.

```rst
**class mod._SchemaMeta(*args: Any, **kwargs: Any)**

**class mod.Schema**

**class mod.Foo**

   **do_something(mod: Schema) -> None**

      :Parameters:
         **schema** (`Schema <#mod.Schema>`_) – Some
         schema.

      :Return type:
         ``None``
```

I've also written pytest cases within the repo without any success. Could you disclose more info? (relevant `conf.py` entry & the rst lines come to my mind). Thanks.

**Comment by bartfeenstra** [comment]
I am getting the following, very similar error:
```
Extension error (sphinx_autodoc_typehints):
Handler <function process_docstring at 0x7e7cf5e92520> for event 'autodoc-process-docstring' threw an exception (exception: unhashable type: 'typing.ParamSpecArgs')
```
This is also since 2.2.0 (pinning to 2.1.1 avoids the bug). I have not been able to determine the exact code sphinx-autodoc-typehints trips over. If and when I do, I'll share an example here to help reproduce the problem.

**Comment by b-phi** [comment]
As requested, some additional info. Will try to find some time to put together a reproducible example

conf.py sections
```python
extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.autosectionlabel',
    'sphinx.ext.napoleon',
    'sphinx_autodoc_typehints',
]

# Autodoc Configuration
autodoc_default_options = {
    'members': True,
    'member-order': 'bysource',
    'special-members': '__init__',
    'undoc-members': True,
}

# Typehint Configuration
typehints_defaults = 'comma'
typehints_use_signature = True
typehints_use_signature_return = True
```
rst where `Foo` is the one in my original comment
```rst
...

.. autoclass:: path.to.class.Foo
```

**Comment by tikuma-lsuhsc** [comment]
@b-phi - Thanks. Those options still won't trigger the error, unfortunately. The same goes with @bartfeenstra's `conf.py` settings in the linked repo.

Meanwhile, I have a fix candidate pushed to my fork:

git+https://github.com/tikuma-lsuhsc/sphinx-autodoc-typehints/tree/fix-issue-462

If possible, could you try the above branch on your projects and see if it resolves the bug?

**Comment by b-phi** [comment]
Hey @tikuma-lsuhsc, unfortunately this branch does not resolve the error

**Comment by tikuma-lsuhsc** [comment]
@bartfeenstra - hmm, that's unfortunate. Regarding your error, do you have a snippet of 2.1.1 output which contains `'typing.ParamSpecArgs'` (or `ParamSpecArgs`)? If so, could you point to the line on your repo as well as the expected output text (html or otherwise)? Thanks

**Comment by b-phi** [comment]
Hey @tikuma-lsuhsc I've created a minimal example. Overriding `__eq__` on the metaclass seems to be required to reproduce for me

```python
from typing import Any


class _SchemaMeta(type):
    def __eq__(cls, other: Any) -> bool:
        return True


class Schema(metaclass=_SchemaMeta):
    pass


def f(s: Schema) -> Schema:
    """
    Do something.

    Args:
        s: Some schema.
    """
    return s

```

**Comment by tikuma-lsuhsc** [comment]
@b-phi - Confirmed! Thanks for putting the example together.

**Comment by tikuma-lsuhsc** [comment]
@b-phi & @bartfeenstra - I think I have the for the issue. If you could please help me test again with my fork (don't forget to update if tried once above):

git+https://github.com/tikuma-lsuhsc/sphinx-autodoc-typehints/tree/fix-issue-462

Thanks!

**Comment by b-phi** [comment]
That seems to have fixed it, thanks!

**Comment by jmccreight** [comment]
This also appears to have fix issues I had (I dont know what they were). Thanks.

**Comment by hoodmane** [comment]
Closed by #463.
